Controller no longer starts up without OAUTH2_ACCESS_TOKEN_URL

The latest PR #108 made `OAUTH2_ACCESS_TOKEN_URL` a required configuration, so it fails to start up (e.g. on demo.zmon.io):

```
Caused by: org.springframework.validation.BindException: org.springframework.boot.bind.RelaxedDataBinder$RelaxedBeanPropertyBindingResult: 1 errors
Field error in object 'tokens' on field 'accessTokenUri': rejected value [${OAUTH2_ACCESS_TOKEN_URL}]; codes [typeMismatch.tokens.accessTokenUri,typeMismatch.accessTokenUri,typeMismatch.java.net.URI,typeMismatch]; arguments [org.springframework.context.support.DefaultMessageSourceResolvable: codes [tokens.accessTokenUri,accessTokenUri]; arguments []; default message [accessTokenUri]]; default message [Failed to convert property value of type [java.lang.String] to required type [java.net.URI] for property 'accessTokenUri'; nested exception is org.springframework.core.convert.ConverterNotFoundException: No converter found capable of converting from type [java.lang.String] to type [java.net.URI]]

--- a/zmon-controller-app/src/main/java/org/kairosdb/client/KairosDBOAuthHttpClient.java
+++ b/zmon-controller-app/src/main/java/org/kairosdb/client/KairosDBOAuthHttpClient.java
@@ -1,0 +1,79 @@
+package org.kairosdb.client;
+
+import org.apache.http.HttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.StringEntity;
+import org.apache.http.impl.client.CloseableHttpClient;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+
+
+/**
+ * OAuth-enabled KairosDB HTTP Client
+ */
+public class KairosDBOAuthHttpClient extends AbstractClient {
+
+    private CloseableHttpClient client;
+    private int retries = 3;
+    private String bearerToken;
+
+    public KairosDBOAuthHttpClient(String url, CloseableHttpClient client, String bearerToken) throws MalformedURLException {
+        super(url);
+        this.client = client;
+        this.bearerToken = bearerToken;
+    }
+
+    @Override
+    protected ClientResponse postData(String json, String url) throws IOException {
+
+        StringEntity requestEntity = new StringEntity(json, ContentType.APPLICATION_JSON);
+        HttpPost postMethod = new HttpPost(url);
+        if (bearerToken != null) {
+            postMethod.setHeader("Authorization", "Bearer " + bearerToken);
+        }
+        postMethod.setEntity(requestEntity);
+
+        return execute(postMethod);
+    }
+
+    @Override
+    protected ClientResponse queryData(String url) throws IOException {
+        throw new UnsupportedOperationException("not implemented");
+    }
+
+    @Override
+    protected ClientResponse delete(String url) throws IOException {
+        throw new UnsupportedOperationException("not implemented");
+    }
+
+    private ClientResponse execute(HttpUriRequest request) throws IOException {
+        HttpResponse response;
+
+        int tries = ++retries;
+        while (true) {
+            tries--;
+            try {
+                response = client.execute(request);
+                break;
+            } catch (IOException e) {
+                if (tries < 1)
+                    throw e;
+            }
+        }
+
+        return new HttpClientResponse(response);
+    }
+
+    @Override
+    public void shutdown() throws IOException {
+        client.close();
+    }
+
+    @Override
+    public int getRetryCount() {
+        return retries;
+    }
+}

--- a/zmon-controller-app/src/main/java/org/zalando/zmon/config/KairosDBProperties.java
+++ b/zmon-controller-app/src/main/java/org/zalando/zmon/config/KairosDBProperties.java
@@ -2,6 +2,7 @@ package org.zalando.zmon.config;
 
 import org.apache.http.client.HttpClient;
 import org.apache.http.client.config.RequestConfig;
+import org.apache.http.impl.client.CloseableHttpClient;
 import org.apache.http.impl.client.HttpClients;
 import org.springframework.boot.context.properties.ConfigurationProperties;
 
@@ -86,7 +87,7 @@ public class KairosDBProperties {
      * get HttpClient with appropriate timeouts
      * @return
      */
-    public HttpClient getHttpClient() {
+    public CloseableHttpClient getHttpClient() {
         RequestConfig config = RequestConfig.custom().setSocketTimeout(getSocketTimeout()).setConnectTimeout(getConnectTimeout()).build();
         return HttpClients.custom().setMaxConnPerRoute(maxConnectionsPerRoute).setMaxConnTotal(maxConnectionsTotal).setDefaultRequestConfig(config).build();
     }

--- a/zmon-controller-app/src/main/java/org/zalando/zmon/controller/KairosDBController.java
+++ b/zmon-controller-app/src/main/java/org/zalando/zmon/controller/KairosDBController.java
@@ -11,7 +11,6 @@ import org.springframework.http.ResponseEntity;
 import org.springframework.util.concurrent.ListenableFuture;
 import org.springframework.util.concurrent.ListenableFutureCallback;
 import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.RequestHeader;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RequestMethod;
 import org.springframework.web.bind.annotation.RestController;
@@ -35,7 +34,7 @@ import com.fasterxml.jackson.databind.node.ObjectNode;
 @RequestMapping(value = "/rest/kairosDBPost")
 public class KairosDBController extends AbstractZMonController {
 
-    private static final String ZMON_DATA_SERVICE_TOKEN_ID = "zmon-data-service";
+    public static final String KAIROSDB_TOKEN_ID = "kairosdb";
 
     private static final String BEARER = "Bearer ";
 
@@ -92,7 +91,7 @@ public class KairosDBController extends AbstractZMonController {
         HttpHeaders headers = new HttpHeaders();
         headers.setContentType(MediaType.APPLICATION_JSON);
         headers.add("X-ZMON-CHECK-ID", checkId);
-        headers.add(AUTHORIZATION, BEARER + accessTokens.get(ZMON_DATA_SERVICE_TOKEN_ID));
+        headers.add(AUTHORIZATION, BEARER + accessTokens.get(KAIROSDB_TOKEN_ID));
 
         HttpEntity<String> httpEntity = new HttpEntity<>(node.toString(), headers);
 
@@ -108,7 +107,7 @@ public class KairosDBController extends AbstractZMonController {
     public ListenableFuture<ResponseEntity<JsonNode>> kairosDBtags(@RequestBody(required = true) final JsonNode node) {
         HttpHeaders headers = new HttpHeaders();
         headers.setContentType(MediaType.APPLICATION_JSON);
-        headers.add(AUTHORIZATION, BEARER + accessTokens.get(ZMON_DATA_SERVICE_TOKEN_ID));
+        headers.add(AUTHORIZATION, BEARER + accessTokens.get(KAIROSDB_TOKEN_ID));
 
         HttpEntity<String> httpEntity = new HttpEntity<>(node.toString(), headers);
 
@@ -121,7 +120,7 @@ public class KairosDBController extends AbstractZMonController {
 
         HttpHeaders headers = new HttpHeaders();
         headers.setContentType(MediaType.APPLICATION_JSON);
-        headers.add(AUTHORIZATION, BEARER + accessTokens.get(ZMON_DATA_SERVICE_TOKEN_ID));
+        headers.add(AUTHORIZATION, BEARER + accessTokens.get(KAIROSDB_TOKEN_ID));
 
         HttpEntity<String> httpEntity = new HttpEntity<>(headers);
 

--- a/zmon-controller-app/src/main/java/org/zalando/zmon/controller/ZMonRestService.java
+++ b/zmon-controller-app/src/main/java/org/zalando/zmon/controller/ZMonRestService.java
@@ -12,7 +12,7 @@ import javax.servlet.http.HttpServletResponse;
 
 import org.apache.http.client.fluent.Executor;
 import org.apache.http.client.fluent.Request;
-import org.kairosdb.client.HttpClient;
+import org.kairosdb.client.KairosDBOAuthHttpClient;
 import org.kairosdb.client.builder.AggregatorFactory;
 import org.kairosdb.client.builder.DataFormatException;
 import org.kairosdb.client.builder.DataPoint;
@@ -37,6 +37,7 @@ import org.springframework.web.bind.annotation.RequestMethod;
 import org.springframework.web.bind.annotation.RequestParam;
 import org.springframework.web.bind.annotation.ResponseBody;
 import org.springframework.web.bind.annotation.ResponseStatus;
+import org.zalando.stups.tokens.AccessTokens;
 import org.zalando.zmon.config.KairosDBProperties;
 import org.zalando.zmon.config.MetricCacheProperties;
 import org.zalando.zmon.domain.CheckDefinition;
@@ -81,6 +82,9 @@ public class ZMonRestService extends AbstractZMonController {
     @Autowired
     ObjectMapper mapper;
 
+    @Autowired
+    AccessTokens accessTokens;
+
     @RequestMapping(value = "/status", method = RequestMethod.GET)
     public ResponseEntity<ExecutionStatus> getStatus() {
         return new ResponseEntity<>(service.getStatus(), HttpStatus.OK);
@@ -100,14 +104,13 @@ public class ZMonRestService extends AbstractZMonController {
         return new ResponseEntity<>(defs, HttpStatus.OK);
     }
 
-    @RequestMapping(value="/updateCheckDefinition")
-    public ResponseEntity<CheckDefinition> updateCheckDefinition(@RequestBody(required=true) CheckDefinitionImport check) {
-        if(check.getOwningTeam() == null || check.getOwningTeam().equals("")) {
-            if(authService.getTeams().isEmpty()) {
+    @RequestMapping(value = "/updateCheckDefinition")
+    public ResponseEntity<CheckDefinition> updateCheckDefinition(@RequestBody(required = true) CheckDefinitionImport check) {
+        if (check.getOwningTeam() == null || check.getOwningTeam().equals("")) {
+            if (authService.getTeams().isEmpty()) {
                 check.setOwningTeam("ZMON");
-            }
-            else {
-                check.setOwningTeam((String)authService.getTeams().toArray()[0]);
+            } else {
+                check.setOwningTeam((String) authService.getTeams().toArray()[0]);
             }
         }
 
@@ -233,7 +236,7 @@ public class ZMonRestService extends AbstractZMonController {
             metric.addAggregator(AggregatorFactory.createAverageAggregator(aggregate, TimeUnit.MINUTES));
         }
 
-        final HttpClient client = new HttpClient(kairosDBProperties.getUrl().toString());
+        final KairosDBOAuthHttpClient client = new KairosDBOAuthHttpClient(kairosDBProperties.getUrl().toString(), kairosDBProperties.getHttpClient(), accessTokens.get(KairosDBController.KAIROSDB_TOKEN_ID));
         try {
             final Long queryStart = System.currentTimeMillis();
             final QueryResponse response = client.query(builder);

--- a/zmon-controller-app/src/main/resources/config/application-zauth.yml
+++ b/zmon-controller-app/src/main/resources/config/application-zauth.yml
@@ -4,3 +4,12 @@ zmon:
   oauth2:
     sso:
       credentials-directory-path: ${CREDENTIALS_DIR}
+
+tokens:
+  access-token-uri: ${OAUTH2_ACCESS_TOKEN_URL}
+  credentials-directory: ${CREDENTIALS_DIR}
+  token-configuration-list:
+    - token-id: kairosdb
+      scopes: uid
+    - token-id: team-service
+      scopes: uid

--- a/zmon-controller-app/src/main/resources/config/application.yml
+++ b/zmon-controller-app/src/main/resources/config/application.yml
@@ -62,13 +62,7 @@ zmon:
           min-idle: 0
           max-wait: -1
 #
-# OAuth2
+# OAuth2: only mock tokens by default
 #
 tokens:
-  access-token-uri: ${OAUTH2_ACCESS_TOKEN_URL}
-  credentials-directory: ${CREDENTIALS_DIR}
-  token-configuration-list:
-    - token-id: zmon-data-service
-      scopes: uid
-    - token-id: team-service
-      scopes: uid
+  enableMock: true

